packages/fuels-code-gen/src/program_bindings/utils.rs

```diff
@@ -1,69 +1,139 @@
+use std::collections::{HashMap, HashSet};
 
 use fuel_abi_types::abi::full_program::FullTypeApplication;
 use inflector::Inflector;
+use itertools::Itertools;
 use proc_macro2::{Ident, TokenStream};
-use quote::{quote, ToTokens};
+use quote::quote;
 
+use super::resolved_type::GenericType;
 use crate::{
     error::Result,
     program_bindings::resolved_type::{ResolvedType, TypeResolver},
-    utils::{safe_ident, TypePath},
+    utils::{self, safe_ident, TypePath},
 };
 
-// Represents a component of either a struct(field name) or an enum(variant
-// name).
 #[derive(Debug)]
-pub(crate) struct Component {
-    pub field_name: Ident,
-    pub field_type: ResolvedType,
+pub(crate) struct Components {
+    components: Vec<(Ident, ResolvedType)>,
 }
 
-impl Component {
+impl Components {
     pub fn new(
-        component: &FullTypeApplication,
+        type_applications: &[FullTypeApplication],
         snake_case: bool,
-        mod_of_component: TypePath,
-    ) -> Result<Component> {
-        let field_name = if snake_case {
-            component.name.to_snake_case()
-        } else {
-            component.name.to_owned()
-        };
-
-        Ok(Component {
-            field_name: safe_ident(&field_name),
-            field_type: TypeResolver::new(mod_of_component).resolve(component)?,
+        parent_module: TypePath,
+    ) -> Result<Self> {
+        let type_resolver = TypeResolver::new(parent_module);
+        let components = type_applications
+            .iter()
+            .map(|type_application| {
+                let name = if snake_case {
+                    type_application.name.to_snake_case()
+                } else {
+                    type_application.name.to_owned()
+                };
+
+                let ident = safe_ident(&name);
+                let ty = type_resolver.resolve(type_application)?;
+                Result::Ok((ident, ty))
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        Ok(Self { components })
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = (&Ident, &ResolvedType)> {
+        self.components.iter().map(|(ident, ty)| (ident, ty))
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.components.is_empty()
+    }
+
+    pub fn as_enum_variants(&self) -> impl Iterator<Item = TokenStream> + '_ {
+        self.components.iter().map(|(ident, ty)| {
+            if let ResolvedType::Unit = ty {
+                quote! {#ident}
+            } else {
+                quote! {#ident(#ty)}
+            }
         })
     }
-}
 
-/// Returns TokenStreams representing calls to `Parameterize::param_type` for
-/// all given Components. Makes sure to properly handle calls when generics are
-/// involved.
-pub(crate) fn param_type_calls(field_entries: &[Component]) -> Vec<TokenStream> {
-    field_entries
-        .iter()
-        .map(|Component { field_type, .. }| single_param_type_call(field_type))
-        .collect()
+    pub fn parameters_for_unused_generics(
```

Comment:
I think adding the prefix `compute_` or `get_` here would make sense, since we're not simply returning a field of the underlying type.

Reply:
Opted for `generate` since the verb is used a lot in the abigen (e.g. `GeneratedCode`, `generate_shared_types` etc.).